File: 9-linked-lists2/2-doubly-linked-lists/main.py
# Doubly Linked List - Python Example

class DNode:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_at_head(self, data):
        new_node = DNode(data)
        new_node.next = self.head
        if self.head:
            self.head.prev = new_node
        self.head = new_node

    def print_forward(self):
        curr = self.head
        while curr:
            print(curr.data, end=" ")
            curr = curr.next
        print()

    def print_backward(self):
        curr = self.head
        if not curr:
            return
        while curr.next:
            curr = curr.next
        while curr:
            print(curr.data, end=" ")
            curr = curr.prev
        print()

dll = DoublyLinkedList()
dll.insert_at_head(3)
dll.insert_at_head(2)
dll.insert_at_head(1)

dll.print_forward()
dll.print_backward()

File: 9-linked-lists2/2-doubly-linked-lists/main.java
// Doubly Linked List - Java Example

class DNode {
    int data;
    DNode prev, next;
    DNode(int data) {
        this.data = data;
        this.prev = this.next = null;
    }
}

class DoublyLinkedList {
    DNode head;

    public void insertAtHead(int data) {
        DNode newNode = new DNode(data);
        newNode.next = head;
        if (head != null) {
            head.prev = newNode;
        }
        head = newNode;
    }

    public void printForward() {
        DNode curr = head;
        while (curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public void printBackward() {
        if (head == null) return;
        DNode curr = head;
        while (curr.next != null) {
            curr = curr.next;
        }
        while (curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.prev;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        DoublyLinkedList dll = new DoublyLinkedList();
        dll.insertAtHead(3);
        dll.insertAtHead(2);
        dll.insertAtHead(1);
        dll.printForward();
        dll.printBackward();
    }
}
